internal/mesh/internal/controllers/sidecarproxy/fetcher/data_fetcher.go

```diff
@@ -0,0 +1,255 @@
+		return nil, nil
+	case err != nil:
+		return nil, err
+	}
+
+	w := &intermediateTypes.Workload{
+		Resource: rsp.Resource,
+	}
+
+	var workload pbcatalog.Workload
+	err = rsp.Resource.Data.UnmarshalTo(&workload)
+	if err != nil {
+		return nil, resource.NewErrDataParse(&workload, err)
+	}
+
+	w.Workload = &workload
+	return w, nil
+}
+
+func (f *Fetcher) FetchProxyStateTemplate(ctx context.Context, id *pbresource.ID) (*intermediateTypes.ProxyStateTemplate, error) {
+	rsp, err := f.Client.Read(ctx, &pbresource.ReadRequest{Id: id})
+
+	switch {
+	case status.Code(err) == codes.NotFound:
+		return nil, nil
+	case err != nil:
+		return nil, err
+	}
+
+	p := &intermediateTypes.ProxyStateTemplate{
+		Resource: rsp.Resource,
+	}
+
+	var tmpl pbmesh.ProxyStateTemplate
+	err = rsp.Resource.Data.UnmarshalTo(&tmpl)
+	if err != nil {
+		return nil, resource.NewErrDataParse(&tmpl, err)
+	}
+
+	p.Tmpl = &tmpl
+	return p, nil
+}
+
+func (f *Fetcher) FetchServiceEndpoints(ctx context.Context, id *pbresource.ID) (*intermediateTypes.ServiceEndpoints, error) {
+	rsp, err := f.Client.Read(ctx, &pbresource.ReadRequest{Id: id})
+
+	switch {
+	case status.Code(err) == codes.NotFound:
+		return nil, nil
+	case err != nil:
+		return nil, err
+	}
+
+	se := &intermediateTypes.ServiceEndpoints{
+		Resource: rsp.Resource,
+	}
+
+	var endpoints pbcatalog.ServiceEndpoints
+	err = rsp.Resource.Data.UnmarshalTo(&endpoints)
+	if err != nil {
+		return nil, resource.NewErrDataParse(&endpoints, err)
+	}
+
+	se.Endpoints = &endpoints
+	return se, nil
+}
+
+func (f *Fetcher) FetchDestinations(ctx context.Context, id *pbresource.ID) (*intermediateTypes.Destinations, error) {
+	rsp, err := f.Client.Read(ctx, &pbresource.ReadRequest{Id: id})
+
+	switch {
+	case status.Code(err) == codes.NotFound:
+		return nil, nil
+	case err != nil:
+		return nil, err
+	}
+
+	u := &intermediateTypes.Destinations{
+		Resource: rsp.Resource,
+	}
+
+	var destinations pbmesh.Upstreams
+	err = rsp.Resource.Data.UnmarshalTo(&destinations)
+	if err != nil {
+		return nil, resource.NewErrDataParse(&destinations, err)
+	}
+
+	u.Destinations = &destinations
+	return u, nil
+}
+
+func (f *Fetcher) FetchDestinationsData(
+	ctx context.Context,
+	destinationRefs []intermediateTypes.CombinedDestinationRef,
+) ([]*intermediateTypes.Destination, map[string]*intermediateTypes.Status, error) {
+
+	var destinations []*intermediateTypes.Destination
+	statuses := make(map[string]*intermediateTypes.Status)
+	for _, dest := range destinationRefs {
+		// Fetch Destinations resource if there is one.
+		us, err := f.FetchDestinations(ctx, dest.ExplicitDestinationsID)
+		if err != nil {
+			// If there's an error, return and force another reconcile instead of computing
+			// partial proxy state.
+			return nil, statuses, err
+		}
+
+		if us == nil {
+			// If the Destinations resource is not found, then we should delete it from cache and continue.
+			f.Cache.DeleteDestination(dest.ServiceRef, dest.Port)
```

Comment:
could you move this into the "not found" handler block of the `FetchDestinations` func as you already did for `FetchWorkload`?

Reply:
Yeah I think I can do that.

Reply:
On a closer look, unfortunately, I cannot because to remove it I need to know both the service ref and the port which I only know from the destinations list passed to this function. If I wanted to do it there, I'd need to fetch all destinations and find the one that's relevant or make cache do it for me. It seems like it's easier to do it here for now.